Implement MemoryStore

DmitryTsepelev opened the pull request:
Here is a very basic implementation of abstract adapter for storages.

The idea is that we keep all the code we had in the Redis adapter and just use the appropriate `store` object for the each specific implementation. This object should be able to respond to all the methods we're using (`get`, `expire at` etc), for the in-memory adapter I've used a `mock_redis` gem (if we don't want to have this dependency - I can replace it with my own implementation). If we once decide to add for instance a postgres support - we'd just need to write the class implementing all the methods we use.

*Note:* I've tried to keep it as much backward compatible as possible but I'm not sure if we really need to configure things like `token_prefix` on the top level but if we want to have it - I'd bring it back.

DmitryTsepelev commented:
Hi @tuwukee, thanks for the review!

I've added a `MemoryStore` (+ tests) implementing all the Redis methods which are in use. Also, I've removed `||=` from adapters and added it to the `JWTSessions#token_store` method to make testing easier.

tuwukee commented on lib/jwt_sessions.rb:
```diff
@@ -89,8 +91,17 @@ def algorithm
+        args
+      elsif args.is_a?(Array)
+        JWTSessions::StoreAdapters::AbstractStoreAdapter.instance_by_symbol(args.first, args.last)
+      end
```
In your tests this method is called as 
```ruby
JWTSessions.store_adapter = :redis
```
Which is an incorrect method usage in this implementation.

DmitryTsepelev replied:
It's not here anymore, probably GH showed one of the previous versions during the outage :)

tuwukee commented on lib/jwt_sessions.rb:
```diff
@@ -89,8 +91,17 @@ def algorithm
+        args
+      elsif args.is_a?(Array)
+        JWTSessions::StoreAdapters::AbstractStoreAdapter.instance_by_symbol(args.first, args.last)
+      end
```
In your tests this method is called as 
```ruby
JWTSessions.store_adapter = :redis
```
Which is an incorrect method usage in this implementation.

tuwukee commented on test/units/jwt_sessions/test_memory_store.rb:
```diff
@@ -0,0 +1,51 @@
+  attr_reader :store
+
+  def setup
+    @store = JWTSessions::StoreAdapters::MemoryStore.new
```
You can set it back to redis on teardown, then there'll be no need to reset it back to redis on setup within other tests.

DmitryTsepelev replied:
Done, I like that idea!

tuwukee commented on test/units/jwt_sessions/test_refresh_token.rb:
```diff
@@ -7,6 +7,7 @@ class TestRefreshToken < Minitest::Test
    attr_reader :csrf, :token, :access_uid
 
   def setup
+    JWTSessions.token_store = :redis
```
As a side note: there're kind of integration tests, defined within sinatra/rails dummy apps.
It'd be great to run them with memory store (and generic unit tests also). 
Can we extend the Rakefile with rake tasks to run the same test suites, but using memory store as default?

DmitryTsepelev replied:
working on it

DmitryTsepelev commented on lib/jwt_sessions.rb:
```diff
@@ -32,13 +36,11 @@ module JWTSessions
   DEFAULT_REDIS_HOST       = '127.0.0.1'
   DEFAULT_REDIS_PORT       = '6379'
   DEFAULT_REDIS_DB_NAME    = '0'
-  DEFAULT_TOKEN_PREFIX     = 'jwt_'
```
👌

DmitryTsepelev commented on lib/jwt_sessions/ext/camelized_string.rb:
```diff
@@ -0,0 +1,11 @@
+
+module JWTSessions
+  module CamelizedString
+    refine String do
```
👍

tuwukee commented on lib/jwt_sessions.rb:
```diff
@@ -89,8 +91,22 @@ def algorithm
   def token_store
-    RedisTokenStore.instance(redis_url, token_prefix)
+    @token_store ||=
+      if ENV['STORE_ADAPTER'].nil?
```
As we're setting the store through ENV for the tests sake only, it'd make sense to set the store in test helpers (or setups) and not in the core code.

DmitryTsepelev replied:
agreed

DmitryTsepelev commented:
@tuwukee I think I've addressed everything we've discussed, please let me know if you see something else, thanks :)

palkan commented on lib/jwt_sessions/store_adapters/abstract_store_adapter.rb:
```diff
@@ -0,0 +1,110 @@
+  module StoreAdapters
+    class AbstractStoreAdapter
+      class << self
+        def instance_by_symbol(adapter, options = {})
```
What about defining registry-related method on `StoreAdapter` module itself? Defining them on abstract adapter class makes things like `RedisStoreAdapter.instance_by_symbol(...)` possible, which looks strange.

DmitryTsepelev replied:
agreed, done

tuwukee commented on test/units/test_jwt_sessions.rb:
```diff
@@ -7,7 +7,6 @@ class TestJWTSessions < Minitest::Test
   def test_default_settings
     assert_equal JWTSessions::DEFAULT_REDIS_HOST, JWTSessions.redis_host
     assert_equal JWTSessions::DEFAULT_REDIS_DB_NAME, JWTSessions.redis_db_name
-    assert_equal JWTSessions::DEFAULT_TOKEN_PREFIX, JWTSessions.token_prefix
```
no need to skip this test

palkan commented on lib/jwt_sessions/store_adapters/abstract_store_adapter.rb:
```diff
@@ -0,0 +1,110 @@
+          Object.const_get(adapter_class_name).instance(options)
+        end
+
+        def instance(token_prefix: nil, **store_options)
```
`instance` implies singleton pattern, which we do not use here (and don't need); let's use a different name, say, `build`

DmitryTsepelev replied:
done

tuwukee commented on Rakefile:
```diff
@@ -16,10 +16,27 @@ RSpec::Core::RakeTask.new(:rails_spec) do |t|
+  :sinatra_spec_with_redis_adapter,
+  :sinatra_spec_with_memory_adapter,
+  :rails_spec_with_redis_adapter,
+  :rails_spec_with_memory_adapter
```
What is the difference between `rails_spec`,  and `rails_spec_with_redis_adapter`/`:rails_spec_with_memory_adapter`?

palkan commented on lib/jwt_sessions/store_adapters/abstract_store_adapter.rb:
```diff
@@ -0,0 +1,110 @@
+          Object.const_get(adapter_class_name).instance(options)
+        end
+
+        def instance(token_prefix: nil, **store_options)
```
We can use default arg value instead of `... || ...` below

```suggestion
        def instance(token_prefix: JWTSessions.token_prefix, **store_options)
```

palkan commented on lib/jwt_sessions/store_adapters/abstract_store_adapter.rb:
```diff
@@ -0,0 +1,110 @@
+          new(@_tokens_store, @_token_prefix)
+        end
+
+        private
```
No need to override `new` to make it private. Just mark it:

```suggestion
        private :new
```

DmitryTsepelev replied:
done

palkan commented on lib/jwt_sessions/store_adapters/abstract_store_adapter.rb:
```diff
@@ -0,0 +1,110 @@
+      end
+
+      def fetch_refresh(uid, namespace)
+        keys   = %i[csrf access_uid access_expiration expiration]
```
Don't we want to move this into a constant and freeze it to avoid allocations?

DmitryTsepelev replied:
done

palkan commented on lib/jwt_sessions/store_adapters/abstract_store_adapter.rb:
```diff
@@ -0,0 +1,110 @@
+        keys.each_with_index.each_with_object({}) { |(key, index), acc| acc[key] = values[index]; }
+      end
+
+      def persist_refresh(uid, access_expiration, access_uid, csrf, expiration, namespace = nil)
```
```suggestion
      def persist_refresh(uid, access_expiration, access_uid, csrf, expiration, namespace = '')
```

?

palkan commented on lib/jwt_sessions/store_adapters/memory_store_adapter.rb:
```diff
@@ -0,0 +1,14 @@
+    class MemoryStoreAdapter < AbstractStoreAdapter
+      class << self
+        def configured_store(options)
+          raise ArgumentError, "unknown keyword: #{options.keys.first}" if options.any?
```
Let's make error message more descriptive, e.g.: "Memory store doesn't support any options"

DmitryTsepelev replied:
👍

palkan commented on lib/jwt_sessions/store_adapters/redis_store_adapter.rb:
```diff
@@ -0,0 +1,16 @@
+  module StoreAdapters
+    class RedisStoreAdapter < AbstractStoreAdapter
+      class << self
+        def configured_store(redis_url: nil)
```
```suggestion
        def configured_store(redis_url: JWTSessions.redis_url)
```

palkan commented on lib/jwt_sessions/store_adapters/abstract_store_adapter.rb:
```diff
@@ -0,0 +1,110 @@
+  module StoreAdapters
+    class AbstractStoreAdapter
+      class << self
+        def instance_by_symbol(adapter, options = {})
```
Maybe, `instance_by_name`? We can use strings too, not only symbols

DmitryTsepelev replied:
👍

palkan commented on lib/jwt_sessions/store_adapters/abstract_store_adapter.rb:
```diff
@@ -0,0 +1,107 @@
+
+module JWTSessions
+  module StoreAdapters
+    def self.build_by_name(adapter, options = {})
```
Let's extract it into `lib/jws_sessions/store_adapters`; this logic doesn't belong to abstract adapter.

We can also move all the `require 'jwt_sessions/store_adapters/*_store'` lines from `jwt_sessions.rb` there.

DmitryTsepelev replied:
👍

palkan commented on lib/jwt_sessions/store_adapters/abstract_store_adapter.rb:
```diff
@@ -0,0 +1,107 @@
+# frozen_string_literal: true
+
+require 'redis'
```
The purpose of abstractizing store adapters is to get rid of required deps.

So, we should not depend on `redis`. It becomes optional.

We should only _require_ it if Redis adapter is used. It's up to user to make sure that `redis` gem is available.

We can make [this trick](https://github.com/rails/rails/blob/master/actioncable/lib/action_cable/subscription_adapter/redis.rb#L5) from Rails to show a meaningful error to user if `redis` gem couldn't be found.

DmitryTsepelev replied:
cool!

palkan commented on lib/jwt_sessions/store_adapters/redis_store_adapter.rb:
```diff
@@ -0,0 +1,16 @@
+    class RedisStoreAdapter < AbstractStoreAdapter
+      class << self
+        def configured_store(redis_url: JWTSessions.redis_url)
+          redis_url ||= JWTSessions.redis_url
```
```suggestion
```

DmitryTsepelev replied:
👍

palkan commented on lib/jwt_sessions/store_adapters/abstract_store_adapter.rb:
```diff
@@ -0,0 +1,107 @@
+      end
+
+      def persist_refresh(uid, access_expiration, access_uid, csrf, expiration, namespace = '')
+        ns = namespace || ''
```
Do we still need this `ns` local variable or we can use `namespace`?

DmitryTsepelev replied:
I've tried to do that but got `TestSession#test_flush_namespaced` failing, because this line makes sure we use an empty string instead of nil even if the nil was explicitly passed (test imitates this exact situation)

palkan commented on lib/jwt_sessions.rb:
```diff
@@ -89,8 +87,19 @@ def algorithm
+
   def token_store
-    RedisTokenStore.instance(redis_url, token_prefix)
+    @token_store ||= StoreAdapters::RedisStoreAdapter.build
```
Using Redis adapter by default along with making it an optional dependency is confusing.

If understand correctly, our goal is to make it possible to use the gem without Redis at all as long as making the upgrade for existent users (who use Redis) as smooth as possible. Right?

I see the following options.

1) Enforce users to specify `token_store` explicitly in the configuration:

```ruby
def token_store
  return @token_store if instance_variable_defined?(:@token_store)
  raise "JWTSessions.token_store is not specified. Please, configure the gem bla-bla-bla"
end
```

2) Check whether `redis` is could be loaded and fallback to in-memory store with warning message:

```ruby
def token_store
  return @token_store if instance_variable_defined?(:@token_store)
  
  begin
    require "redis"
    self.token_store = :redis
  rescue LoadError
    warn <<~MSG
      Warning! JWTSessions uses in-memory token store.
      Since you haven't configured JWTSessions.token_store, we tried to use Redis but couldn't load the 'redis' gem.
      
      Make sure that 'redis' gem is present in your Gemfile (or configure a different store).
    MSG
    self.token_store = :memory
  end
end
```

@tuwukee @DmitryTsepelev WDYT?

DmitryTsepelev replied:
I like the second option because it won't affect the existing users who use redis in their apps while the first option will force them to make an additional configuration

DmitryTsepelev commented:
One more topic to discuss - in the current implementation we have an `AbstractStoreAdapter` with it's subclasses. All the common logic about working with storage in encapsulated inside of it so it sends messages to the `store` object. This object is configured by child classes, and frankly saying it's the only one responsibility subclasses have.

`store` can either contain a Redis client or a memory storage, which implements all the Redis methods we use. Technically these two could have a parent abstract class, and if one day we decide to add a PostgreSQL adapter we would have to find out what methods should be implemented by looking into the memory store, add the story and one more adapter (right, with one method with store configuration). Also, if we decide to use one more Redis method - it's not super clear that we should implement something else in our custom adapters.

Summing up all the issues with the current approach:

1. Concrete adapters just confuse store
2. No explicit interface for store
3. Adding new adapter would require adding two classes - adapter and store

I'm thinking about the following solution: instead of calling methods on `store` our `AbstractStoreAdapter` should define it's own methods related to manipulations with store and leave them unimplemented, while concrete adapters should configure stores in a way they want and implement these methods. For instance, instead of `store.expireat(key, timestamp) we could call `set_expiration_for(key, timestamp)`, redis adapter would simply delegate everything to Redis client, while memory store would contain everything we have in the `MemoryStore` right now. As a result - clear interface and no parallel hierarchies (aka _bridge_).

@tuwukee @palkan what do you think?

palkan commented:
> I'm thinking about the following solution: instead of calling methods on store our AbstractStoreAdapter should define it's own methods related to manipulations with store and leave them unimplemented, while concrete adapters should configure stores in a way they want and implement these methods. 

+1
Makes sense.

tuwukee commented:
Cool, removal of unnecessary abstractions is a nice idea.

DmitryTsepelev commented:
Done! Also, I've added some notes about configuring store adapter to the readme

tuwukee commented on README.md:
```diff
@@ -281,9 +281,11 @@ end
 
-Default token store configurations
+In order to configure token store you should either set up a concrete instance of store adapter (`JWTSessions.token_store = JWTSessions::StoreAdapters::RedisStoreAdapter.build(redis_url: 'redis://127.0.0.1:6379/0')`) or pass a symbol and optional options (`JWTSessions.token_store = :redis, { redis_url: 'redis://127.0.0.1:6379/0', token_prefix: 'prefix' }`). Currently supported stores are `:redis` ...
+
```
Besides `redis_url` there're also such settings as `redis_host`, `redis_port`, `redis_db_name`. 
Taking all that into account it feels like the configuration becomes super confusing.

Is it just me, or a basic setting notation like this
```ruby
JWTSessions.token_store = JWTSessions::StoreAdapters::RedisStoreAdapter.build(redis_url: 'redis://127.0.0.1:6379/0')`
```
looks overwhelming?
Let's keep only `JWTSessions.token_store = :redis, options`, what do you think?

Also it's super minor, but... "optional options".

DmitryTsepelev replied:
> Besides redis_url there're also such settings as redis_host, redis_port, redis_db_name. Taking all that into account it feels like the configuration becomes super confusing.

Yeah, let's make the adapter to accept these options too, but keep backward compatibility as usual

> Let's keep only JWTSessions.token_store = :redis, options, what do you think?
Yeah, I like that, not sure if anyone would prefer writing the exact classes

and yes, optional options🤦‍♂️

DmitryTsepelev replied:
addressed all the mentioned issues

tuwukee commented on lib/jwt_sessions.rb:
```diff
@@ -89,8 +87,34 @@ def algorithm
+          Since you haven't configured JWTSessions.token_store, we tried to use Redis but couldn't load the 'redis' gem.
+
+          Make sure that 'redis' gem is present in your Gemfile (or configure a different store).
+        MSG
```
It looks almost like we're blaming users for not having redis installed :D
Let's add a more polite message, something like
```
Warning! JWTSessions uses in-memory token store.
Unless token store is not specified explicitly, JWTSessions uses Redis by default and fallbacks to in-memory token store.

Make sure that 'redis' gem is present in your Gemfile (or configure a different store).
```

DmitryTsepelev replied:
yep, changed the wording, thanks :)

DmitryTsepelev commented on lib/jwt_sessions.rb:
```diff
@@ -89,8 +78,30 @@ def algorithm
     @algorithm ||= DEFAULT_ALGORITHM
   end
 
+  def token_store=(args)
```
when someone calls `JWTSessions.token_store = :redis, { ... }` ruby passes two args as a single array, so with the version above we'd have `store = [:redis, { ... }]` and `options = {}`

tuwukee replied:
you're right, carry on.

DmitryTsepelev commented:
@tuwukee @palkan  I think I've addressed everything we've discussed so far, so I've squashed everything into a single commit. Let me know if you see something I could improve, thanks!

DmitryTsepelev commented:
As a followup for https://github.com/tuwukee/jwt_sessions/pull/39#issuecomment-433054867, I've redesigned API for adapters.

Now each descendant has to implement the following public methods:

- fetch_access
- persist_access
- fetch_refresh
- persist_refresh
- update_refresh
- all_in_namespace
- destroy_refresh
- destroy_access

Also, the following protected methods should be implemented too:

- configure_store (I can't imagine an adapter which does not need a post-initialization hook `configure_store`)
- wildcard_refresh_key (used in `refresh_key`)

palkan commented:
> I can't imagine an adapter which does not need a post-initialization hook configure_store

Why can't we use `#initialize`? (and let adapter decide whether it needs `prefix` or not, for example, in-memory adapter doesn't need it, does it?)

DmitryTsepelev commented:
Okay, `prefix` is left read-only at the top level and concrete adapters can define it if needed, all the configuration happens inside the `initialize`. Also, I've removed the part that made `new` private - it's not a public API anymore, and the user has to use symbols anyway.

tuwukee commented on lib/jwt_sessions/store_adapters/redis_store_adapter.rb:
```diff
@@ -0,0 +1,105 @@
+    class RedisStoreAdapter < AbstractStoreAdapter
+      attr_reader :redis_client
+
+      def initialize(token_prefix: JWTSessions.token_prefix, **options)
```
Looks like this implementation won't pick up settings specified as follows
```
JWTSessions.redis_port = 6378
JWTSessions.token_store = :redis
```

DmitryTsepelev replied:
https://github.com/tuwukee/jwt_sessions/pull/39/files#diff-f9f1ac5ee448bd383409144652d60a5dR93 is responsible for that - it takes all the passed args and after that makes a `||=` fallback to the module-level config options. I've added a test to verify that

tuwukee commented on test/units/test_token_store.rb:
```diff
@@ -0,0 +1,47 @@
+    JWTSessions.redis_port = 6378
+    JWTSessions.token_store = :redis
+
+    assert_equal 'redis://127.0.0.1:6378/0', JWTSessions.token_store.redis_client.connection[:id]
```
redis port should be reset on the teardown

DmitryTsepelev replied:
Good catch, fixed that

batizhevsky commented on test/units/jwt_sessions/store_adapters/test_memory_store_adapter.rb:
```diff
@@ -0,0 +1,68 @@
+    store.persist_refresh('uid', expiration, 'access_uid', 'csrf', expiration, 'ns')
+    store.persist_refresh('uid', expiration, 'access_uid', 'csrf', expiration, 'ns2')
+    refresh_tokens = store.all('ns')
+    assert_equal refresh_tokens.count, 1
```
I suppose arguments should be in the different order:
`assert_equal(exp, act, msg = nil)`

DmitryTsepelev replied:
agreed, let me check all other specs

DmitryTsepelev commented:
Rebased once again, I guess it's ready to be merged unless there are any other issues you see :)
